File: LeetCode/BST/python/getMinDifference.py
# Given the root of a Binary Search Tree (BST), return the minimum absolute difference between the values of any two different nodes in the tree.

 

# Example 1:


# Input: root = [4,2,6,1,3]
# Output: 1
# Example 2:


# Input: root = [1,0,48,null,null,12,49]
# Output: 1
 

# Constraints:

# The number of nodes in the tree is in the range [2, 104].
# 0 <= Node.val <= 105
 

# Note: This question is the same as 783: https://leetcode.com/problems/minimum-distance-between-bst-nodes/

# TC: O(n)
# SC: O(h), h is the height of the tree

# Definition for a binary tree node.
from collections import deque
from typing import Dict, List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:

    def __init__(self):
        self.min_distance = [float('inf')]
        self.prev = [None]

    def getMinimumDifference(self, root: Optional[TreeNode]) -> int:
        # Binary Search Tree

        # Call recursive DFS
        self.dfs(root)

        return self.min_distance[0]
    
    def dfs(self, node: Optional[TreeNode]) -> None:
        if node is None:
            return

        self.dfs(node.left)

        if self.prev[0] is not None:
            self.min_distance[0] = min(self.min_distance[0], node.val - self.prev[0])

        self.prev[0] = node.val

        self.dfs(node.right)
    
    @staticmethod
    def testSolution(input: Dict[str, List[str|int]]) -> None:
        bTree = Solution.BFSCreateBinaryTree(input["root"])
        print(f"Input: {Solution.BFSBinaryTreeToStr(bTree)}")
        print(f"Expected: {input['expected']}")

        res = Solution().getMinimumDifference(bTree)

        print(f"Result: {res}")
        print("\033[92mPASS\033[00m" if res == input.get("expected", [[]]) else "\033[91mFAIL\033[00m")

    @staticmethod
    def BFSCreateBinaryTree(node: List[int|str]) -> TreeNode:
        arLen = len(node)

        if arLen < 1:
            return None

        root = TreeNode(node[0], None, None)
        i = 1

        # Using queue (BFS) to build the tree level by level
        queue = deque([root])

        while len(queue) > 0 and i < arLen:
            # Pre order: root, left, then right
            cur_node = queue.popleft()

            # Left child
            if i < arLen and node[i] != "null":
                cur_node.left = TreeNode(node[i], None, None)
                queue.append(cur_node.left)
            i += 1
            
            # Right child
            if i < arLen and node[i] != "null":
                cur_node.right = TreeNode(node[i], None, None)
                queue.append(cur_node.right)            
            i += 1

        return root
    
    @staticmethod
    def BFSBinaryTreeToStr(root: Optional[TreeNode]) -> None:

        if root is None:
            return "[]"
        
        sb = []

        # Using queue (BFS) to build the tree level by level
        queue = deque([root])


        while len(queue) > 0:
            curNode = queue.popleft()

            tmp = str(curNode.val) if curNode is not None else "null"

            sb.append(tmp)
 
            if curNode is not None:
                queue.append(curNode.left)
                queue.append(curNode.right)

        new_len = 0

        # Check starting from end of sb_array where the value is not "null"
        for i in range(len(sb) - 1, -1, -1):
            if sb[i] == "null":
                continue
            else:
                new_len = i
                break

        # Remove those extra "null"
        if len(sb) > 0:
            sb = sb[0 : new_len + 1]

        return "[" + ", ".join(sb) + "]"


if __name__ == "__main__":
    # Example test cases
    test_cases = [
        {"root": [4,2,6,1,3], "expected": 1},
        {"root": [1,0,48,"null","null",12,49], "expected": 1},
    ]

    for i, test_case in enumerate(test_cases):
        print(f"Test Case {i+1}:")
        Solution.testSolution(test_case)
        print(f"{'-'*30}")

File: LeetCode/BST/JavaScript/getMinDifference.js
// Given the root of a Binary Search Tree (BST), return the minimum absolute difference between the values of any two different nodes in the tree.

// Example 1:

// Input: root = [4,2,6,1,3]
// Output: 1
// Example 2:

// Input: root = [1,0,48,null,null,12,49]
// Output: 1

// Constraints:

// The number of nodes in the tree is in the range [2, 104].
// 0 <= Node.val <= 105

// Note: This question is the same as 783: https://leetcode.com/problems/minimum-distance-between-bst-nodes/

const Result = { PASS: "\x1b[92mPASS\x1b[0m", FAIL: "\x1b[91mFAIL\x1b[0m" };

class Record {
  constructor(root, expected) {
    this.root = root;
    this.expected = expected;
  }
}

class TreeNode {
  constructor(val, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

class Solution {
  /**
   * @param {TreeNode} root
   * @return {number}
   */
  getMinimumDifference(root) {
    let min_dist = [parseFloat(Number.POSITIVE_INFINITY)];
    let prev = [null];

    function dfs(node) {
      if (node === null) {
        return;
      }

      // In order Traversal

      dfs(node.left);

      // Update the new min_distance
      if (prev[0] !== null) {
        min_dist[0] = Math.min(min_dist[0], node.val - prev[0]);
      }

      // Update the prev value
      prev[0] = node.val;

      dfs(node.right);
    }

    dfs(root);

    return min_dist[0];
  }
}

function buildTree(values) {
  if (values.length === 0 || values[0] === null) {
    return null;
  }

  const root = new TreeNode(values[0]);
  const queue = [root];
  let index = 1;

  while (queue.length > 0 && index < values.length) {
    const node = queue.shift();

    if (values[index] !== null) {
      node.left = new TreeNode(values[index]);
      queue.push(node.left);
    }
    index++;

    if (index < values.length && values[index] !== null) {
      node.right = new TreeNode(values[index]);
      queue.push(node.right);
    }
    index++;
  }

  return root;
}

function testSolution(record) {
  console.log(`input:\troot: ${JSON.stringify(record.root)}`);
  console.log(`expected: ${record.expected}`);

  const solution = new Solution();
  const result = solution.getMinimumDifference(buildTree(record.root));

  console.log(`result: ${result}`);
  console.log(result === record.expected ? Result.PASS : Result.FAIL);
}

const records = [
  new Record([4, 2, 6, 1, 3], 1),
  new Record([1, 0, 48, null, null, 12, 49], 1),
];

records.forEach((record, i) => {
  console.log(`# Test case ${i + 1}`);
  testSolution(record);
  console.log("-".repeat(50));
});
